File: typescript/packages/bench/src/support/abstractService.ts
/** Base shared by all 12 services. */

import { v4 as uuid } from "uuid";

import type { AuditEntry, AuditableContract } from "../contracts/auditableContract";
import { lastAuditOf, type HasAudit } from "../concerns/hasAudit";

export abstract class AbstractService implements AuditableContract, HasAudit {
    private readonly trail: string[] = [];

    protected constructor(readonly name: string) {}

    /** Record one action on this service's trail. */
    record(action: string): void {
        this.trail.push(action);
    }

    auditTrail(): readonly string[] {
        return this.trail;
    }

    lastAudit(): string | undefined {
        return lastAuditOf(this.trail);
    }

    auditDepth(): number {
        return this.trail.length;
    }

    auditActor(): string {
        return this.name;
    }

    audit(action: string): AuditEntry {
        return { id: uuid(), actor: this.auditActor(), action };
    }
}

File: rust/crates/atelier-bench/src/support/abstract_service.rs
//! Base shared by all 12 services.

use crate::concerns::has_audit::HasAudit;
use crate::contracts::auditable_contract::AuditableContract;

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct ServiceBase {
    pub name: &'static str,
    trail: Vec<String>,
}

impl ServiceBase {
    pub fn new(name: &'static str) -> Self {
        Self {
            name,
            trail: Vec::new(),
        }
    }

    /// Record one action on this service's trail.
    pub fn record(&mut self, action: impl Into<String>) {
        self.trail.push(action.into());
    }
}

impl HasAudit for ServiceBase {
    fn audit_trail(&self) -> &[String] {
        &self.trail
    }
}

impl AuditableContract for ServiceBase {
    fn audit_actor(&self) -> String {
        self.name.to_string()
    }
}
